Don't allow setting the ElkLoggerConfigProvider directly (breaking change)

Remove the setter method for `ElkLoggerConfigProvider`.

Yes, according to semantic versioning, this _should_ require a major version bump.

But I'm making an exception here, because sometimes rules should be broken.

The real reason for making the exception is that there is no way in "real life" to actually do this. `ElkLoggerConfigProvider` doesn't have a public no-args constructor, and it isn't a typical Dropwizard configuration class, so there's no way you can set it in YAML anyway. Plus, applications do not directly instantiate `ElkAppenderFactory`.

So, trying to do something like the following makes no sense and doesn't work:

```yaml
logging:
  level: INFO
  loggers:
    integration-test: DEBUG
    org.kiwiproject.elk: DEBUG
  appenders:
    - type: console
    - type: elk
      elkLoggerConfigProvider:
        host: dev-logstash-1.acme.com
        port: 7005
```

It blows up in rather spectacular fashion:

```
  * Failed to parse configuration at: logging.appenders.[1].elkLoggerConfigProvider; Cannot construct instance of `org.kiwiproject.config.provider.ElkLoggerConfigProvider` (no Creators, like default constructor, exist): cannot deserialize from Object value (no delegate- or property-based Creator)
 at [Source: UNKNOWN; byte offset: #UNKNOWN] (through reference chain: org.kiwiproject.elk.DropwizardTestAppExtension$TestAppConfiguration["logging"]->io.dropwizard.logging.common.DefaultLoggingFactory["appenders"]->java.util.ArrayList[1]->org.kiwiproject.elk.ElkAppenderFactory["elkLoggerConfigProvider"])

        at io.dropwizard.configuration.ConfigurationParsingException$Builder.build(ConfigurationParsingException.java:277)
        at io.dropwizard.configuration.BaseConfigurationFactory.build(BaseConfigurationFactory.java:177)
        at io.dropwizard.configuration.BaseConfigurationFactory.build(BaseConfigurationFactory.java:94)
        at io.dropwizard.core.cli.ConfiguredCommand.parseConfiguration(ConfiguredCommand.java:139)
        at io.dropwizard.core.cli.ConfiguredCommand.run(ConfiguredCommand.java:85)
        at io.dropwizard.testing.DropwizardTestSupport.startIfRequired(DropwizardTestSupport.java:304)
        at io.dropwizard.testing.DropwizardTestSupport.before(DropwizardTestSupport.java:204)
        at io.dropwizard.testing.junit5.DropwizardAppExtension.before(DropwizardAppExtension.java:186)
        at io.dropwizard.testing.junit5.DropwizardAppExtension.beforeAll(DropwizardAppExtension.java:175)
        at org.kiwiproject.elk.DropwizardTestAppExtension.beforeAll(DropwizardTestAppExtension.java:55)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `org.kiwiproject.config.provider.ElkLoggerConfigProvider` (no Creators, like default constructor, exist): cannot deserialize from Object value (no delegate- or property-based Creator)
 at [Source: UNKNOWN; byte offset: #UNKNOWN] (through reference chain: org.kiwiproject.elk.DropwizardTestAppExtension$TestAppConfiguration["logging"]->io.dropwizard.logging.common.DefaultLoggingFactory["appenders"]->java.util.ArrayList[1]->org.kiwiproject.elk.ElkAppenderFactory["elkLoggerConfigProvider"])
        at com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:67)
        at com.fasterxml.jackson.databind.DeserializationContext.reportBadDefinition(DeserializationContext.java:1943)
        at com.fasterxml.jackson.databind.DatabindContext.reportBadDefinition(DatabindContext.java:415)
        at com.fasterxml.jackson.databind.DeserializationContext.handleMissingInstantiator(DeserializationContext.java:1430)
        at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1514)
        at com.fasterxml.jackson.module.blackbird.deser.SuperSonicBeanDeserializer.deserializeFromObject(SuperSonicBeanDeserializer.java:247)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)
        at com.fasterxml.jackson.module.blackbird.deser.SuperSonicBeanDeserializer.deserialize(SuperSonicBeanDeserializer.java:116)
        at com.fasterxml.jackson.module.blackbird.deser.SettableObjectProperty.deserializeAndSet(SettableObjectProperty.java:44)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:302)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:207)
        at com.fasterxml.jackson.module.blackbird.deser.SuperSonicBeanDeserializer.deserialize(SuperSonicBeanDeserializer.java:120)
        at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedForId(AsPropertyTypeDeserializer.java:169)
        at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:135)
        at com.fasterxml.jackson.databind.deser.AbstractDeserializer.deserializeWithType(AbstractDeserializer.java:262)
        at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer._deserializeNoNullChecks(CollectionDeserializer.java:503)
        at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer._deserializeFromArray(CollectionDeserializer.java:358)
        at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:245)
        at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:29)
        at com.fasterxml.jackson.module.blackbird.deser.SettableObjectProperty.deserializeAndSet(SettableObjectProperty.java:44)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:302)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:207)
        at com.fasterxml.jackson.module.blackbird.deser.SuperSonicBeanDeserializer.deserialize(SuperSonicBeanDeserializer.java:120)
        at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedUsingDefaultImpl(AsPropertyTypeDeserializer.java:225)
        at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:144)
        at com.fasterxml.jackson.databind.deser.AbstractDeserializer.deserializeWithType(AbstractDeserializer.java:262)
        at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:138)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:265)
        at com.fasterxml.jackson.module.blackbird.deser.SuperSonicBeanDeserializer.deserialize(SuperSonicBeanDeserializer.java:155)
        at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:342)
        at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:5011)
        at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3104)
        at io.dropwizard.configuration.BaseConfigurationFactory.build(BaseConfigurationFactory.java:148)
        ... 9 more
```

As a result, I am just removing that setter method entirely, since it cannot be used in any normal Dropwizard application. And if by some extremely remote chance someone found a way to use this class programmatically and _is_ using that setter method, _lo siento_.

--- a/src/main/java/org/kiwiproject/elk/ElkAppenderFactory.java
+++ b/src/main/java/org/kiwiproject/elk/ElkAppenderFactory.java
@@ -12,6 +12,8 @@ import io.dropwizard.logging.common.AbstractAppenderFactory;
 import io.dropwizard.logging.common.async.AsyncAppenderFactory;
 import io.dropwizard.logging.common.filter.LevelFilterFactory;
 import io.dropwizard.logging.common.layout.LayoutFactory;
+import lombok.AccessLevel;
+import lombok.Getter;
 import lombok.Setter;
 import net.logstash.logback.appender.LogstashTcpSocketAppender;
 import net.logstash.logback.appender.LogstashUdpSocketAppender;
@@ -24,20 +26,107 @@ import org.kiwiproject.json.JsonHelper;
 import java.util.HashMap;
 import java.util.Map;
 
+/**
+ * An implementation of {@link io.dropwizard.logging.common.AppenderFactory AppenderFactory}
+ * that sends log messages to Logstash (the L in the ELK stack).
+ * <p>
+ * You don't use this directly, but instead configure it in your Dropwizard configuration file.
+ * For example, here is a configuration that uses only default values:
+ * <pre>
+ * logging:
+ *   level: WARN
+ *   loggers:
+ *     org.acme.service: INFO
+ *   appenders:
+ *     - type: elk
+ * </pre>
+ * And here's an example that uses UDP, doesn't include the logger context, and overrides
+ * some of the default Logstash field names:
+ * <pre>
+ * logging:
+ *   level: WARN
+ *   loggers:
+ *     org.acme.service: INFO
+ *   appenders:
+ *     - type: elk
+ *       useUdp: true
+ *       includeContext: false
+ *       fieldNames:
+ *         logger: loggerName
+ *         thread: threadName
+ *         level: logLevel
+ * </pre>
+ *  The available configuration properties are listed below.
+ * <table>
+ *     <caption>Configuration properties</caption>
+ *     <tr>
+ *         <td>Name</td>
+ *         <td>Default</td>
+ *         <td>Description</td>
+ *     </tr>
+ *     <tr>
+ *         <td>{@code useUdp}</td>
+ *         <td>{@code false}</td>
+ *         <td>Whether to use UDP for connections to Logstash.</td>
+ *     </tr>
+ *     <tr>
+ *         <td>{@code includeCallerData}</td>
+ *         <td>{@code false}</td>
+ *         <td>
+ *             Whether to include caller data, required for line numbers.
+ *             Beware, this is expensive as it creates a new Throwable on each logging call.
+ *         </td>
+ *     </tr>
+ *     <tr>
+ *         <td>{@code includeContext}</td>
+ *         <td>{@code true}</td>
+ *         <td>Whether to include the logging context in log messages.</td>
+ *     </tr>
+ *     <tr>
+ *         <td>{@code includeMdc}</td>
+ *         <td>{@code true}</td>
+ *         <td>Whether to include the MDC (Message Diagnostic Context) in log messages.</td>
+ *     </tr>
+ *     <tr>
+ *         <td>{@code customFields}</td>
+ *         <td>empty map</td>
+ *         <td>Custom fields that will be included in log messages.</td>
+ *     </tr>
+ *     <tr>
+ *         <td>{@code fieldNames}</td>
+ *         <td>empty map</td>
+ *         <td>
+ *             Map containing Logstash field name mappings.
+ *             These override the default values in {@link net.logstash.logback.fieldnames.LogstashFieldNames LogstashFieldNames}.
+ *         </td>
+ *     </tr>
+ * </table>
+ */
 @Setter
+@Getter(AccessLevel.PACKAGE)  // getters are visible with package scope for testing
 @JsonTypeName("elk")
 public class ElkAppenderFactory extends AbstractAppenderFactory<ILoggingEvent> {
 
     private static final JsonHelper JSON_HELPER = JsonHelper.newDropwizardJsonHelper();
 
     private boolean useUdp;
-    private boolean includeCallerData;
-    private boolean includeContext = true;
-    private boolean includeMdc = true;
-    private Map<String, String> customFields = new HashMap<>();
-    private Map<String, String> fieldNames = new HashMap<>();
+    @Getter(AccessLevel.PUBLIC) private boolean includeCallerData;  // must be public; getter in superclass is public
+    private boolean includeContext;
+    private boolean includeMdc;
+    private Map<String, String> customFields;
+    private Map<String, String> fieldNames;
+    @Setter(AccessLevel.NONE) private ElkLoggerConfigProvider elkLoggerConfigProvider;  // don't allow setting this
 
-    private ElkLoggerConfigProvider elkLoggerConfigProvider = ElkLoggerConfigProvider.builder().build();
+    /**
+     * Create a new instance with default values.
+     */
+    public ElkAppenderFactory() {
+        includeContext = true;
+        includeMdc = true;
+        customFields = new HashMap<>();
+        fieldNames = new HashMap<>();
+        elkLoggerConfigProvider = ElkLoggerConfigProvider.builder().build();
+    }
 
     @Override
     public Appender<ILoggingEvent> build(LoggerContext loggerContext,

--- a/src/test/java/org/kiwiproject/elk/ElkAppenderFactoryTest.java
+++ b/src/test/java/org/kiwiproject/elk/ElkAppenderFactoryTest.java
@@ -2,6 +2,7 @@ package org.kiwiproject.elk;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.junit.jupiter.api.Assertions.assertAll;
 
 import ch.qos.logback.classic.LoggerContext;
 import ch.qos.logback.classic.spi.ILoggingEvent;
@@ -37,6 +38,25 @@ class ElkAppenderFactoryTest {
         loggerContext = new LoggerContext();
         filterFactory = new ThresholdLevelFilterFactory();
         appenderFactory = new AsyncLoggingEventAppenderFactory();
+    }
+
+    @Nested
+    class Constructor {
+
+        @Test
+        void shouldSetDefaultValues() {
+            var factory = new ElkAppenderFactory();
+
+            assertAll(
+                () -> assertThat(factory.isUseUdp()).isFalse(),
+                () -> assertThat(factory.isIncludeCallerData()).isFalse(),
+                () -> assertThat(factory.isIncludeContext()).isTrue(),
+                () -> assertThat(factory.isIncludeMdc()).isTrue(),
+                () -> assertThat(factory.getCustomFields()).isEmpty(),
+                () -> assertThat(factory.getFieldNames()).isEmpty(),
+                () -> assertThat(factory.getElkLoggerConfigProvider()).isNotNull()
+            );
+        }
     }
 
     @Nested

